src/xtgeo/common/sys.py

```diff
@@ -99,126 +143,126 @@ def generic_hash(gid, hashmethod="md5"):
+
+        if not isinstance(filelike, (str, pathlib.Path, io.BytesIO, io.StringIO)):
+            raise RuntimeError(
+                f"Cannot instantiate {self.__class__} from "
+                f"{filelike} of type {type(filelike)}. Expected "
+                f"a str, pathlib.Path, io.BytesIO, or io.StringIO."
+            )
+
+        self._file: Union[pathlib.Path, io.BytesIO, io.StringIO]
+        self._tmpfile: Optional[str] = None
         self._delete_after = False  # delete file (e.g. tmp) afterwards
         self._mode = mode
-        self._memstream = False
 
         self._cfhandle = 0
         self._cfhandlecount = 0
 
         # for internal usage in tests; mimick window/mac with no fmemopen in C
         self._fake_nofmem = False
 
-        logger.debug("Init ran for _XTGeoFile")
+        if isinstance(filelike, str):
+            filelike = pathlib.Path(filelike)
 
-        # The self._file must be a Pathlib or a BytesIO instance
-        if isinstance(fobj, pathlib.Path):
-            self._file = fobj
-        elif isinstance(fobj, str):
-            self._file = pathlib.Path(fobj)
-        elif isinstance(fobj, io.BytesIO):
-            self._file = fobj
-            self._memstream = True
-        elif isinstance(fobj, io.StringIO):
-            self._file = fobj
-            self._memstream = True
-        elif isinstance(fobj, _XTGeoFile):
-            raise RuntimeError("Reinstancing object, not allowed", self.__class__)
-        else:
-            raise RuntimeError(
-                f"Illegal input, cannot continue ({self.__class__}) "
-                f"{fobj}: {type(fobj)}"
-            )
+        self._file = filelike
+        self._memstream = isinstance(self._file, (io.BytesIO, io.StringIO))
 
         if obj and not self._memstream:
             self.resolve_alias(obj)
 
-        logger.info("Ran init of %s, ID is %s", __name__, id(self))
+        logger.debug("Ran init of %s, ID is %s", __name__, id(self))
 
     @property
-    def memstream(self):
-        """Read only: Get True if file object is a memory stream (BytesIO)."""
+    def memstream(self) -> bool:
+        """Get whether or not this file is a io.BytesIO/StringIO memory stream."""
         return self._memstream
 
     @property
-    def file(self):
-        """Read only: Get Path object (if input was file) or BytesIO object."""
+    def file(self) -> Union[pathlib.Path, io.BytesIO, io.StringIO]:
+        """Get Path object (if input was file) or memory stream object."""
         return self._file
 
     @property
-    def name(self):
-        """The absolute path name of a file."""
-        logger.info("Get absolute name of file...")
-
-        if self._memstream:
-            return self._file
+    def name(self) -> Union[str, io.BytesIO, io.StringIO]:
+        """Get the absolute path name of the file, or the memory stream."""
+        if isinstance(self.file, (io.BytesIO, io.StringIO)):
+            return self.file
```

Comment:
There is a lot of this type narrowing around checking for memory streams in this PR. mypy doesn't seem sophisticated enough to know that `self.memstream` will be True when the filelike it initialized with is a StringIO/BytesIO. I tried a few different interations of getting around this like doing an explicit `isinstance` check in the property method but no dice.

Very interested in suggestions! But I think it is a mypy limitation.